### PR TITLE
Ged walk centrality

### PR DESCRIPTION
**Pull request by angriman**
This PR brings the implementation of the algorithms for GedWalk centrality based on the paper [Group Centrality Maximization for Large-scale Graphs](https://arxiv.org/abs/1910.13874) (ALENEX 2020).

Updating tlx is needed because the algorithms use some features introduced in a newer tlx version.

### CODE REVIEW

File: include/networkit/centrality/GedWalk.hpp [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+#include <networkit/graph/Graph.hpp>
+
+#include <tlx/container/d_ary_addressable_int_heap.hpp>
+
+namespace NetworKit {
+
+class GedWalk final : public Algorithm {
+
+    using walks = double;
+    using uchar = unsigned char;
```
Comment (avdgrinten) [inline include/networkit/centrality/GedWalk.hpp]: Non-idiomatic typedef. Please remove this.

File: include/networkit/centrality/GedWalk.hpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    };
+
+    struct EvaluationResult {
+        double score;
+        walks w;
+    };
+
+    tlx::d_ary_addressable_int_heap<node, 2, CompareScore> scoreQ{CompareScore(gainScore)};
+    tlx::d_ary_addressable_int_heap<node, 2, CompareScore> boundQ{CompareScore(gainBound)};
+    CSRMatrix gMat;
```
Comment (avdgrinten) [inline include/networkit/centrality/GedWalk.hpp]: I think it makes more sense to use a graph here. Yes, a CSR will give slightly better performance, but it requires an entire copy of the graph.

File: include/networkit/centrality/GedWalk.hpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    std::vector<walks> gainW;      // Number of marginal walks on the last level.
+    std::vector<double> gainBound; // Upper bound on the marginal score.
+
+    struct CompareScore {
+    public:
+        CompareScore(const std::vector<double> &score) : score(score) {}
+        bool operator()(node x, node y) const { return score[x] > score[y]; }
+
+    private:
+        const std::vector<double> &score;
```
Comment (avdgrinten) [inline include/networkit/centrality/GedWalk.hpp]: Member variables should not be references.

File: include/networkit/centrality/GedWalk.hpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    EvaluationResult evaluateGraph();
+    EvaluationResult evaluateGroup();
+
+    template <typename T>
+    void parallelFill(std::vector<T> &vec, const T &val) {
+#pragma omp parallel for
+        for (omp_index i = 0; i < static_cast<omp_index>(vec.size()); ++i) {
+            vec[i] = val;
+        }
+    }
```
Comment (avdgrinten) [inline include/networkit/centrality/GedWalk.hpp]: Does this actually improve performance? I would expect that there is too little work done in this loop to make parallelization worthwhile (over auto-vectorization / lowering to `memset()` which should happen anyway).

It adds some maintainance cost as there is mental overhead to remember what this function does (compared to the open-coded `for` loop that ones expects for this purpose), so maybe it would be better to open-code it.

File: include/networkit/centrality/GedWalk.hpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * Return the GedWalk score of the input group.
+     *
+     * @param first/last The range that contains the input group.
+     * @param scoreEpsilon The precision of the score to be computed.
+     */
+    template <class InputIt>
+    double scoreOfGroup(InputIt first, InputIt last, double scoreEpsilon = .1);
```
Comment (avdgrinten) [inline include/networkit/centrality/GedWalk.hpp]: Is this named consistently with the other group centralities?

File: include/networkit/centrality/GedWalk.hpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+     * Run the algorithm.
+     */
+    void run() override;
+
+    /**
+     * Return the approximate GedWalk score of the computed group.
+     *
+     * @return The approximate score of the computed group.
+     */
+    double getApxScore() const {
```
Comment (avdgrinten) [inline include/networkit/centrality/GedWalk.hpp]: This name is confusing.

File: include/networkit/centrality/GedWalk.hpp [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+     */
+    template <class InputIt>
+    double scoreOfGroup(InputIt first, InputIt last, double scoreEpsilon = .1);
+
+    /**
+     * Return the computed group.
+     *
+     * @return The computed group.
+     */
+    std::vector<node> groupMaxGedWalk() const {
```
Comment (avdgrinten) [inline include/networkit/centrality/GedWalk.hpp]: Is this named consistently with the other group centralities?

File: networkit/cpp/centrality/test/CentralityGTest.cpp [hunk 8]
```diff
@@ -244,8 +245,8 @@ TEST_F(CentralityGTest, testKatzDynamicAddition) {
     DEBUG("finish kc");
     node u, v;
     do {
-        u = GraphTools::randomNode(G);
-        v = GraphTools::randomNode(G);
+        u = G.randomNode();
+        v = G.randomNode();
```
Comment (avdgrinten) [inline networkit/cpp/centrality/test/CentralityGTest.cpp]: Random unrelated changes to the PR. Please remove this.

File: include/networkit/centrality/GedWalk.hpp (hunk 7 above)
Comment (angriman) [reply include/networkit/centrality/GedWalk.hpp]: Yes, but we might want to change this in the future if we want to build an abstract GroupCentrality class.

File: include/networkit/centrality/GedWalk.hpp (hunk 5 above)
Comment (angriman) [reply include/networkit/centrality/GedWalk.hpp]: Yes

File: include/networkit/centrality/GedWalk.hpp (hunk 1 above)
Comment (angriman) [reply include/networkit/centrality/GedWalk.hpp]: I removed it.

File: include/networkit/centrality/GedWalk.hpp (hunk 2 above)
Comment (angriman) [reply include/networkit/centrality/GedWalk.hpp]: Alright now GedWalk does not use an additional CSR matrix anymore.

File: include/networkit/centrality/GedWalk.hpp (hunk 4 above)
Comment (angriman) [reply include/networkit/centrality/GedWalk.hpp]: Removed it in favor of std::fill

File: include/networkit/centrality/GedWalk.hpp (hunk 6 above)
Comment (angriman) [reply include/networkit/centrality/GedWalk.hpp]: Renamed to `getApproximateScore`

File: networkit/cpp/centrality/test/CentralityGTest.cpp (hunk 8 above)
Comment (angriman) [reply networkit/cpp/centrality/test/CentralityGTest.cpp]: That was a mistake, I removed this change.

**Comment by angriman** [comment]
Thanks @avdgrinten for the review, I've addressed all the comments.

File: include/networkit/centrality/GedWalk.hpp [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        return group;
+    }
+
+    const std::vector<double> &boundOfMarginalGains() const { return gainBound; }
+};
+
+template <class InputIt>
+double GedWalk::scoreOfGroup(InputIt first, InputIt last, double scoreEpsilon) {
+    if (boundStrategy == BoundStrategy::spectral) {
+        assert(alpha < 1 / static_cast<double>(sigmaMax));
```
Comment (fabratu) [inline include/networkit/centrality/GedWalk.hpp:160]: Only a very small remark: Depending on the value of `sigmaMax` and `degInMax`, the assertion can benefit (both in precision and clocktime) if compare `alpha * static_cast<double>(something)` to `1`.
